src/diffusers/models/unet_2d_blocks.py

```diff
@@ -2198,3 +2465,215 @@ def forward(
                 hidden_states = upsampler(hidden_states, temb)
 
         return hidden_states
+
+
+class KUpBlock2D(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        temb_channels: int,
+        dropout: float = 0.0,
+        num_layers: int = 5,
+        resnet_eps: float = 1e-5,
+        resnet_act_fn: str = "gelu",
+        resnet_groups: Optional[int] = None,
+        resnet_group_size: Optional[int] = 32,
+        add_upsample=True,
+        is_first_block: bool = False,
+        resnet_time_scale_shift: str = "scale_shift",
+    ):
+        super().__init__()
+        resnets = []
+        k_in_channels = out_channels if is_first_block else 2 * out_channels
+        k_mid_channels = out_channels
+        k_out_channels = in_channels
+        num_layers = num_layers - 1
+
+        for i in range(num_layers):
+            resnets.append(
+                ResnetBlock2D(
+                    in_channels=k_in_channels if i == 0 else k_mid_channels,
+                    mid_channels=k_mid_channels,
+                    out_channels=k_out_channels if (i == num_layers - 1) else k_mid_channels,
+                    temb_channels=temb_channels,
+                    eps=resnet_eps,
+                    groups=resnet_groups,
+                    group_size=resnet_group_size,
+                    dropout=dropout,
+                    non_linearity=resnet_act_fn,
+                    norm1_scale_shift=True,
+                    time_embedding_norm=resnet_time_scale_shift,
+                    dropout_after_conv=True,
+                    conv_shortcut_bias=False,
+                )
+            )
+
+        self.resnets = nn.ModuleList(resnets)
+
+        if add_upsample:
+            self.upsamplers = nn.ModuleList([KUpsample2d()])
+        else:
+            self.upsamplers = None
+
+        self.gradient_checkpointing = False
+
+    def forward(self, hidden_states, res_hidden_states_tuple, temb=None, upsample_size=None):
+        res_hidden_states_tuple = res_hidden_states_tuple[-1]
```

Comment:
@yiyixuxu note how the blocks again absorb the complexity so that `Unet2DConditionModel` can stay simple